compiler/optimizer/abstractinterpreter/AbsValue.hpp

```diff
@@ -0,0 +1,136 @@
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at http://eclipse.org/legal/epl-2.0
+ * or the Apache License, Version 2.0 which accompanies this distribution
+ * and is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the
+ * Eclipse Public License, v. 2.0 are satisfied: GNU General Public License,
+ * version 2 with the GNU Classpath Exception [1] and GNU General Public
+ * License, version 2 with the OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#ifndef ABS_VALUE_INCL
+#define ABS_VALUE_INCL
+
+#include "il/OMRDataTypes.hpp"
+#include "compile/Compilation.hpp"
+#include "optimizer/VPConstraint.hpp"
+#include "optimizer/ValuePropagation.hpp"
+
+namespace TR { class AbsVPValue; }
+
+namespace TR {
+
+/**
+ * AbsValue is the abstract representation of a 'value'.
+ * It is the basic unit used to perform abstract interpretation.
+ * This class is an abstract class.
+ */
+class AbsValue
+   {
+   public:
+   AbsValue(TR::DataType dataType) :
+         _dataType(dataType),
+         _paramPos(-1),
+         _isImplicitParameter(false)
+   {}
+
+   /**
+    * @brief Clone an abstract value
+    * 
+    * @param region The region where the cloned value will be allocated on.
+    * @return the cloned abstract value
+    */
+   virtual TR::AbsValue* clone(TR::Region& region)=0;
+
+   /**
+    * @brief Merge with another AbsValue. 
+    * @note This is an in-place merge. Other should have the exactly same dataType as self. 
+    *
+    * @param other Another AbsValue to be merged with
+    * @return NULL if failing to merge. Self if succeding to merge.
+    */
+   virtual TR::AbsValue* merge(TR::AbsValue *other)=0;
+
+   /**
+    * @brief Check whether the AbsValue is least precise abstract value.
+    * @note Top denotes the least precise representation in lattice theory.
+    *
+    * @return true if it is top. false otherwise
+    */
+   virtual bool isTop()=0;
+
+   /**
+    * @brief Set to the least precise abstract value.
+    */
+   virtual void setToTop()=0;
+
+   /**
+    * @brief Check if the AbsValue is a parameter.
+    *
+    * @return true if it is a pramater. false if not.
+    */
+   bool isParameter() { return _paramPos >= 0; }
+
+   /**
+    * @brief Check if the AbsValue is 'this' (the implicit parameter.)
+    *
+    * @return true if it is the implicit parameter. false otherwise.
+    */   
+   bool isImplicitParameter() { return _paramPos == 0 && _isImplicitParameter; }
```

Comment:
Why is this information needed? What is the use-case for knowing whether a value is an "implicit parameter"? What does it mean to be an "implicit parameter" in the general case? The description given of an "implicit parameter" being the `this` argument is specific to Java and this PR seems to lack a language-agnostic definition.

Reply:
This is useful during abstract interpretation. For Java, if a parameter is an implicit parameter, we know it can never be NULL, so there cannot be a NULLCHK on this parameter. Yes, "this" is a little misguiding and I will remove it.

Reply:
I don't think this is the right way to represent such information. Why not just represent the non-nullness as part of the constraint internally instead?

Reply:
No, this is not how non-nullness is used during abstract interpretation. If we know a parameter is non-null, we know if inlining this method, the null-check on this parameter can be eliminated. There would be a null-check folding optimization.

But for the implicit parameter, for example, in Java, there is not even an implicit null check on it. So during abstract interpretation, we would not consider the implicit parameter has a chance to have null check eliminated since there is no null check. Only knowing the value of being non-null is not enough.

Reply:
I think this flag can be removed since we can check if a method is static or not to tell if a parameter is implicit or not.